Initial introduction to SLSA docs (#11)
enterprise-contract/ec-cookbook: modules/ROOT/pages/slsa.adoc

= Red Hat Trusted Application Pipeline, Enterprise Contract, and SLSA

.Overview
****

Red Hat Trusted Application Pipeline (RHTAP) is an industry leading cloud build

> system that aims to meet link:https://slsa.dev/spec/v1.0/levels[SLSA Build L3].

Reviewer: Is it worth calling out that we're targeting Build L3? or could we keep it more generic and say we're looking to meet or exceed the SLSA spec. By linking directly to Build L3, that ties to a specific definition which is true today, but may not be true later, and remembering to come back and update this link if that changes seems ambitious.

Author: I used that wording since I thought it was one of the marketing points. I used the softer "aims to meet" rather than "meets" in case we aren't quite sure yet. Are you suggesting something like "aims to meet or exceed SLSA Build L3"? Or something else?

signature can't be verified for any reason, a failure will be reported. There are other ways to verify image signatures, e.g. by using `cosign` directly,
but EC will happily do it for you.

There is also experimental support in EC for validating signatures for images signed keylessly
using https://docs.sigstore.dev/fulcio/overview/[Sigstore Fulcio], though RHTAP currently uses long-lived
signing keys to sign images and attestations.

a| https://slsa.dev/spec/v1.0/verifying-artifacts[Verifying Artifacts (v1.0)]
a| Attestation signature check passed
a| As well as verifying the image signature, EC also verifies the signature of the attestation, which is an important part of the artifact
verification process. For RHTAP the same key is used to sign the attestation and the image itself. As mentioned above, there is experimental
support for verifying keylessly signed attestations.

a| https://slsa.dev/spec/v1.0/verifying-artifacts#step-1-check-slsa-build-level[Verifying Artifacts (v1.0)]
a| xref:ec-policies:ROOT:release_policy.adoc#slsa_build_scripted_build__subject_build_task_matches[Provenance subject matches build task image result]

> a| SLSA recommends an artifact verifier checks that the that provenance statement’s subject matches the digest of the artifact in question.

Reviewer: Extra words here "checks that the that provenance..."

> This is a good sanity check to confirm that the provenance being considered is for expected image. Because EC uses Cosign under the hood, and

Reviewer: ```suggestion
is for the expected image.
```

Reviewer: It may be worth considering if there's a different way to phrase "sanity check" here. IBM's style guide forbids using that phrase in any Red Hat documentation. Not sure if that applies to upstream documentation too. See https://issues.redhat.com/browse/STONEINTG-233 for when the Integration team renamed their tasks.

Author: Ack, thanks. Changed it to "performs and additional confirmation".

Cosign does this check as part of its signature verification, we get this confirmed automatically. The EC rule linked here does an additional
sanity check by comparing the image reference produced by the build task and confirming that also matches the provenance subject.

a| https://slsa.dev/spec/v1.0/verifying-artifacts#step-1-check-slsa-build-level[Verifying Artifacts (v1.0)]
a| xref:ec-policies:ROOT:release_policy.adoc#slsa_provenance_available__attestation_predicate_type_accepted[Expected attestation predicate type found]
a| SLSA suggests that the predicate type in the artifact's provenance should be `https://slsa.dev/provenance/v1`. This rule confirms that requirement,

> and will fail if the predicate type found to be any other value.

Reviewer: ```suggestion
and will fail if the predicate type is found to be any other value.
```

Author: :heavy_check_mark:

a| This is a good sanity check related to artifact verification that isn't specifically mentioned in the SLSA guidelines. In RHTAP build pipelines the
git clone task provides information about the git repo and the commit sha used. This rule confirms that the url and the commit sha from the git clone
task match what appears in the "materials" provenance.

|===

== Notes and caveats

Because EC was designed initially for use with RHTAP, and RHTAP uses Tekton
CI/CD, many of the specific rules mentioned above are currently Tekton specific
in that they use details about the Tekton pipeline run, which are currently
made available in the SLSA attestation created by Tekton Chains.

This is not a limitation of EC itself, but rather just an attribute of the
xref:ec-policies:ROOT:release_policy.adoc[policies created for RHTAP] in their
current state. In future we're aiming to extract the Tekton specific rules from
the general SLSA rules to make the rules more generally useful. It's expected
that policy rules for build systems not based on Tekton and Tekton Chains could
be created easily.

> Note also that since EC can conveniently apply any policy defined in OPA/Rego,

Reviewer: Would it be worth embedding links for OPA or Rego to take them somewhere that describes those things?

it is able to be used for a wide range of purposes, e.g. to verify certain tests
were run and passed, and to apply any business specific release policies.

Browse the rules defined in xref:ec-policies:ROOT:release_policy.adoc[Policies] to
see what other rules have been defined for use by RHTAP within Red Hat.

== Conclusion

Enterprise Contract can be used as a convenient and customizable SLSA artifact
verifier. It's doing that now within RHTAP, and we think it's capable and
flexible enough to be adapted to other build systems.

If you have questions, or want to provide feedback, please reach out to the

> Enterprise Contract team via https://github.com/enterprise-contract[GitHub].

Reviewer: How should they reach out? By opening an issue? On which repo?

Author: Not sure really, but I guess we'll have this info available on the org page.

Author: Readme at https://github.com/enterprise-contract says file an issue, I think it's good enough for now.

customizable policy checker designed to be easily integrated with RHTAP's CI/CD
workflows.

A lot of the default policies created for RHTAP and enforced by EC are closely
related to SLSA requirements, based on both the recent SLSA v1.0, and the older
SLSA v0.1 specification. In this documentation we'll describe in more detail
how RHTAP and EC work together to provide a high standard of supply chain
security, informed and guided by the link:https://slsa.dev/[SLSA
specifications].

****

== SLSA Requirements

=== SLSA requirements verified explicitly by Enterprise Contract

Enterprise Contract comes with a set of policy rules designed for use with
RHTAP. Using EC to verify that a build passes all the checks defined by the
policy provides a high level of confidence that the build was built securely
and can be trusted.

Many of the policy rules are linked to specific SLSA requirements, for example:

[cols="1,1,3"]
|===
|SLSA reference|EC policy rule|Notes

a| https://slsa.dev/spec/v0.1/requirements#scripted-build[Scripted build (v0.1)]
a| xref:ec-policies:ROOT:release_policy.adoc#slsa_build_scripted_build__build_script_used[Build task contains steps]

> a| RHTAP uses link:https://tekton.dev/[Tekton] so knowing that is probably enough to consider this requirement satisfied, however this specific

Reviewer 2: ```suggestion
a| RHTAP uses link:https://tekton.dev/[Tekton] so knowing that is enough to consider this requirement satisfied, however this specific
```

Reviewer 2: Just, stronger language.

It will leave a better impression since this is the _first_ table row that the reader encounters.

can be used to verify the authenticity of the signed provenances.

a| https://slsa.dev/spec/v1.0/levels#build-l3-hardened-builds[Build L3: Hardened builds]
a| For example, the RHTAP's Tekton environment prevents different build pipelines from influencing each
other, and has secure access polices and management practices for signing secrets and other credentials.

|===

////
In future there may be some RHTAP system level docs addressing this in more detail.
Linking to a general system overview of RHTAP would also be useful here also I think.
////

== Notes and caveats

Because EC was designed initially for use with RHTAP, and RHTAP uses Tekton
CI/CD, many of the specific rules mentioned above are currently Tekton specific
in that they use details about the Tekton pipeline run, which are currently
made available in the SLSA attestation created by Tekton Chains.

This is not a limitation of EC itself, but rather just an attribute of the
xref:ec-policies:ROOT:release_policy.adoc[policies created for RHTAP] in their
current state. In future we're aiming to extract the Tekton specific rules from

> the general SLSA rules to make the rules more generally useful. It's expected

Reviewer 2: I dunno. Just so as not to imply that they're not useful atm:

```suggestion
the general SLSA rules to make the rules more generally applicable. It's expected
```

Reviewer 2: Just to match the speaker of the sentence before.

```suggestion
the general SLSA rules to make the rules more generally useful. We expect
```

Author: I missed the second one intially, but added in commit 2af8b8e .
